backend/director/core/reasoning.py

```
@@ -117,12 +118,25 @@ def register_agents(self, agents: List[BaseAgent]):
 
+    def _trim_context(self, max_messages: int) -> None:
+        """Trim reasoning_context to at most max_messages, always preserving index 0 (system prompt)."""
+        ctx = self.session.reasoning_context
+        if len(ctx) <= max_messages:
+            return
+        trimmed = len(ctx) - max_messages
+        logger.warning(
+            f"Context window trimmed: dropping {trimmed} oldest message(s) "
+            f"({len(ctx)} → {max_messages})"
+        )
+        self.session.reasoning_context = [ctx[0]] + ctx[-(max_messages - 1):]
```

Ruff (0.15.10)

[warning] Consider `[ctx[0], *ctx[-(max_messages - 1):]]` instead of concatenation

Replace with `[ctx[0], *ctx[-(max_messages - 1):]]`

(RUF005)